proposals/fibers/Explainer.md

```diff
@@ -0,0 +1,874 @@
+
+#### `fiber.switch` Switch to a different fiber
+
+The `fiber.switch` instruction is a combination of a `fiber.suspend` and a `fiber.resume` to an identified fiber. This instruction is useful for circumstances where the suspending fiber knows which other fiber should be resumed.
+
+The `fiber.switch` instruction has three arguments: the identity of the fiber being suspended, the identity of the fiber being resumed and the signaling event.
+
+Although it may be viewed as being a combination of the two instructions, there is an important distinction also: the signaling event. Under the common hierarchical organization, a suspending fiber does not know which fiber will be resumed. This means that the signaling event has to be of a form that the fiber's manager is ready to process. However, with a `fiber.switch` instruction, the fiber's manager is not informed of the switch and does not need to understand the signaling event.
+
+This, in turn, means that a fiber manager may be relieved of the burden of communicating between fibers. I.e., `fiber.switch` supports a symmetric coroutining pattern. However, precisely because the fiber's manager is not made aware of the switch between fibers, it must also be the case that this does not _matter_; in effect, the fiber manager may not directly be aware of any of the fibers that it is managing.  
+
+#### `fiber.retire` Retire a fiber
+
+The `fiber.retire` instruction is used when a fiber has finished its work and wishes to inform its parent of any final results. Like `fiber.suspend` (and `fiber.resume`), `fiber.retire` has an event argument&mdash;together with associated values on the agument stack&mdash; that are communicated.
+
+In addition, the retiring fiber is put into a moribund state and any computation resources associated with it are released. If the fiber has any active descendants then they too are made moribund.
+
+>It is not recommended that a fiber allows exceptions to be propagated out of the fiber function. Instead, the function should use a `fiber.retire` &mdash;together with an appropriate event description&mdash;to signal the exceptional return. This allows the resume ancestor to directly capture the exceptional event as part of its normal response to the resume.
+
+>The reason that we don't recommend allowing exceptions to propagate is that an inapprpriate exception handler may be invoked as a result. This is especially dangerous in the case that the retiring fiber was switched to&mdash;with a `fiber.switch` instruction&mdash;rather than being resumed.
+
+#### `fiber.retireto` Retire a fiber and directly switch
```

Comment:
The only difference between this and doing a `fiber.switch` followed by a normal return is that the fiber is marked moribund and its resources freed sooner. But given that it would be a toolchain bug if a moribund fiber were resumed, the only real benefit here is the eager resource freeing. It seems that this instruction could be replaced with a `fiber.switch` + `fiber.release`.

Reply:
Again, a key semantic difference is that retireto does not involve the scheduler. Also, it is not generally replaceable by switch+release because that would require that the target fiber 'knows' that the 'sending' fiber is done. Again, that is not always possible or appropriate.

Reply:
Note that not involving a separate scheduler is (a) something that is a tail_call-like optimization for some languages and (b) imperative when communicating data between sibling fibers. The latter scenario allows unboxed values to be used in more situations.